File: backend/src/main/java/rra/orbitcontrol/services/ProductoService.java
package rra.orbitcontrol.services;

import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Sort;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import rra.orbitcontrol.config.PushNotificationService;
import rra.orbitcontrol.exceptions.EntityNotFoundException;
import rra.orbitcontrol.mappers.ProductoMapper;
import rra.orbitcontrol.models.dtos.responses.ProductoDTO;
import rra.orbitcontrol.models.entities.Producto;
import rra.orbitcontrol.models.enums.RolNombre;
import rra.orbitcontrol.models.enums.TipoProducto;
import rra.orbitcontrol.repositories.ProductoRepository;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

@Service
public class ProductoService {

    private final ProductoRepository      productoRepository;
    private final ProductoMapper          productoMapper;
    private final PushNotificationService pushNotificationService;

    public ProductoService(ProductoRepository productoRepository, ProductoMapper productoMapper,
                           PushNotificationService pushNotificationService) {
        this.productoRepository      = productoRepository;
        this.productoMapper          = productoMapper;
        this.pushNotificationService = pushNotificationService;
    }

    public List<ProductoDTO> all() {
        return productoMapper.toDtoList(productoRepository.findAllWithProveedor());
    }

    // Combina búsqueda de texto y filtro por tipo de producto; si vienen los dos aplica los dos a la vez
    public Map<String, Object> allPaged(int pagina, int tamanio, String sortby, String sortDir, String search, String filtro) {
        Sort sort = sortDir.equalsIgnoreCase("asc") ? Sort.by(sortby).ascending() : Sort.by(sortby).descending();
        PageRequest pageable = PageRequest.of(pagina, tamanio, sort);

        boolean hasSearch = search != null && !search.isBlank();
        boolean hasFiltro = filtro != null && !filtro.isBlank();

        Page<Producto> page;
        if (hasFiltro) {
            TipoProducto tipo = TipoProducto.valueOf(filtro);
            page = hasSearch
                ? productoRepository.searchPagedByTipo(search, tipo, pageable)
                : productoRepository.findByTipoProducto(tipo, pageable);
        } else {
            page = hasSearch ? productoRepository.searchPaged(search, pageable) : productoRepository.findAll(pageable);
        }

        Map<String, Object> response = new HashMap<>();
        response.put("productos",    productoMapper.toDtoList(page.getContent()));
        response.put("currentPage",  page.getNumber());
        response.put("totalPages",   page.getTotalPages());
        response.put("total",        page.getTotalElements());
        return response;
    }

    public List<ProductoDTO> findByTipo(TipoProducto tipo) {
        return productoMapper.toDtoList(productoRepository.findByTipoProducto(tipo));
    }

    public Map<String, Object> stockBajo(int pagina, int tamanio) {
        PageRequest pageable = PageRequest.of(pagina, tamanio, Sort.by("stockActual").ascending());
        Page<Producto> page  = productoRepository.findStockBajo(pageable);
        Map<String, Object> response = new HashMap<>();
        response.put("productos",   productoMapper.toDtoList(page.getContent()));
        response.put("currentPage", page.getNumber());
        response.put("totalPages",  page.getTotalPages());
        response.put("total",       page.getTotalElements());
        return response;
    }

    public void enviarAlerta(Long id) {
        Producto p = productoRepository.findById(id)
                .orElseThrow(() -> new EntityNotFoundException("Producto no encontrado: " + id));
        pushNotificationService.enviarARoles(
            List.of(RolNombre.ENCARGADO, RolNombre.ADMIN),
            "Stock bajo — " + p.getNombre(),
            "Quedan " + p.getStockActual() + ". Mínimo: " + p.getStockMinimo(),
            "/almacen/alertas"
        );
    }

    // Envía push notification por cada producto con stock bajo (máximo 100 a la vez)
    public void enviarTodasAlertas() {
        PageRequest pageable = PageRequest.of(0, 100, Sort.by("stockActual").ascending());
        productoRepository.findStockBajo(pageable).getContent().forEach(p ->
            pushNotificationService.enviarARoles(
                List.of(RolNombre.ENCARGADO, RolNombre.ADMIN),
                "Stock bajo — " + p.getNombre(),
                "Quedan " + p.getStockActual() + ". Mínimo: " + p.getStockMinimo(),
                "/almacen/alertas"
            )
        );
    }

    @Transactional
    public ProductoDTO save(Producto producto) {
        Producto saved = productoRepository.save(producto);
        // Recargamos con JOIN para que el DTO incluya los datos del proveedor,
        // que la entidad recién guardada no tiene populados todavía
        Producto conProveedor = productoRepository.findByIdWithProveedor(saved.getId())
                .orElse(saved);
        return productoMapper.toDto(conProveedor);
    }

    public ProductoDTO one(Long id) {
        Producto producto = productoRepository.findByIdWithProveedor(id)
                .orElseThrow(() -> new EntityNotFoundException("producto", id));
        return productoMapper.toDto(producto);
    }

    @Transactional
    public ProductoDTO replace(Long id, Producto producto) {
        productoRepository.findById(id)
                .orElseThrow(() -> new EntityNotFoundException("producto", id));
        producto.setId(id);
        Producto saved = productoRepository.save(producto);
        Producto conProveedor = productoRepository.findByIdWithProveedor(saved.getId())
                .orElse(saved);
        return productoMapper.toDto(conProveedor);
    }

    @Transactional
    public void delete(Long id) {
        if (!productoRepository.existsById(id)) {
            throw new EntityNotFoundException("producto", id);
        }
        productoRepository.deleteById(id);
    }
}

File: frontend/src/app/core/services/producto.service.ts
import { Injectable, inject } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { map, Observable } from 'rxjs';
import { Producto, TipoProducto } from '../models/producto.model';

export interface PageResponse<T> {
  productos: T[];
  currentPage: number;
  totalPages: number;
  total: number;
}

@Injectable({
  providedIn: 'root',
})

export class ProductoService {
  private http = inject(HttpClient);
  private readonly API = '/api/productos';

  getAll(): Observable<Producto[]> {
    return this.http.get<Producto[]>(this.API);
  }

  getAllPaged(pagina: number, tamanio: number = 10, sortBy = 'nombre', sortDir = 'asc', search = '', filtro = ''): Observable<PageResponse<Producto>> {
    return this.http.get<PageResponse<Producto>>(this.API, {
      params: { pagina: pagina.toString(), tamanio: tamanio.toString(), sortBy, sortDir, search, filtro }
    });
  }

  getById(id: number): Observable<Producto> {
    return this.http.get<Producto>(`${this.API}/${id}`);
  }

  getStockBajo(pagina: number = 0, tamanio: number = 5): Observable<Producto[]> {
      return this.http.get<PageResponse<Producto>>(`${this.API}/alertas/stockBajo`, {
          params: { pagina: pagina.toString(), tamanio: tamanio.toString() }
      }).pipe(
          map(response => response.productos)
      );
  }

  getByTipo(tipoProducto: TipoProducto): Observable<Producto[]> {
    return this.http.get<Producto[]>(`${this.API}/tipoProducto/${tipoProducto}`);
  }

  create(producto: Producto): Observable<Producto> {
    return this.http.post<Producto>(this.API, producto);
  }

  update(id: number, producto: Producto): Observable<Producto> {
    return this.http.put<Producto>(`${this.API}/${id}`, producto);
  }

  delete(id: number): Observable<void> {
    return this.http.delete<void>(`${this.API}/${id}`);
  }

  enviarAlerta(id: number): Observable<void> {
    return this.http.post<void>(`${this.API}/${id}/alerta`, {});
  }

  enviarTodasAlertas(): Observable<void> {
    return this.http.post<void>(`${this.API}/alertas/enviar`, {});
  }
}
